RFC for DCR relocation to CSNX (#164)
guardian/csnx: docs/project-imports/dcr.md

# Importing `dotcom-rendering` into `csnx`

## Glossary

- `dotcom-rendering` refers to the application that renders HTML for `frontend`
- _"the `dotcom-rendering` repo"_ refers to the repo that contains `dotcom-rendering`, `apps-rendering` and `common-rendering` codebases

## Intention

We want to relocate `dotcom-rendering` in the `csnx` monorepo.

### Why?

- to make it easier for Guardian product teams to ship by reducing the friction of interdependency between `dotcom-rendering` and other Guardian projects (Source, Libs, consent, commercial code etc)
- to reduce the time spent by dotcom on managing their infrastructure
- to standardise process: _best practice by default_

## Preparation

### Turn the `dotcom-rendering` repo into a single `yarn` workspace

the `dotcom-rendering` repo is currently managed by two different package managers:

1. A `yarn` workspace containing `dotcom-rendering` and `common-rendering`
2. An `npm`-managed package that contains the `apps-rendering` code.

#### Problem

- `apps-rendering` consumes some undeclared dependencies from the `yarn` workspace

> - some `dotcom-rendering`-specific dependencies are installed in the root

Reviewer: ```suggestion
- some `dotcom-rendering`-specific dependencies are installed in the root
- it impossible to know for certain which packages are installed by which project and where, which makes it impossibly to confidently remove part of the workspace
```

# Importing `dotcom-rendering` into `csnx`

## Glossary

- `dotcom-rendering` refers to the application that renders HTML for `frontend`
- _"the `dotcom-rendering` repo"_ refers to the repo that contains `dotcom-rendering`, `apps-rendering` and `common-rendering` codebases

## Intention

We want to relocate `dotcom-rendering` in the `csnx` monorepo.

### Why?

- to make it easier for Guardian product teams to ship by reducing the friction of interdependency between `dotcom-rendering` and other Guardian projects (Source, Libs, consent, commercial code etc)
- to reduce the time spent by dotcom on managing their infrastructure
- to standardise process: _best practice by default_

## Preparation

### Turn the `dotcom-rendering` repo into a single `yarn` workspace

the `dotcom-rendering` repo is currently managed by two different package managers:

1. A `yarn` workspace containing `dotcom-rendering` and `common-rendering`
2. An `npm`-managed package that contains the `apps-rendering` code.

> 

Reviewer: ```suggestion
```

#### Problem

- `apps-rendering` consumes some undeclared dependencies from the `yarn` workspace
- some `dotcom-rendering`-specific dependencies are installed in the root
- it impossible to know for certain which packages are installed by which project and where, which makes it impossibly to confidently remove part of the workspace

#### Solution

- convert `apps-rendering` to `yarn`
- move all projects into a single `yarn` workspace
- move all possible deps to individual project's `package.json`

### Add Nx to the `dotcom-rendering` repo for use in `dotcom-rendering`

#### Problem

- `csnx` uses Nx to orchestrate the project
- we want the actual importing of `dotcom-rendering` code to require as few changes as possible

#### Solution

- we will add Nx to the existing the `dotcom-rendering` repo monorepo
- the setup will be as close as possible to `csnx`.

_n.b. `apps-rendering` and `common-rendering` will both be able to use Nx, although setting this up will not be CSTI's initial focus._

### Remove unused config from the `dotcom-rendering` repo

#### Problem

> - there is a lot of config files in the `dotcom-rendering` repo

Reviewer: ```suggestion
- there are a lot of config files in the `dotcom-rendering` repo
```

# Importing `dotcom-rendering` into `csnx`

## Glossary

- `dotcom-rendering` refers to the application that renders HTML for `frontend`
- _"the `dotcom-rendering` repo"_ refers to the repo that contains `dotcom-rendering`, `apps-rendering` and `common-rendering` codebases

## Intention

We want to relocate `dotcom-rendering` in the `csnx` monorepo.

### Why?

> - to make it easier for Guardian product teams to ship by reducing the friction of interdependency between `dotcom-rendering` and other Guardian projects (Source, Libs, consent, commercial code etc)

Reviewer: potentially, definitely. i don't think we have opinion yet on what happens after DCR but that's more the absence of a plan, rather than a plan to _not_ move things.

we also want to make it easier for everyone to understand how/what relocating projects needs so it's not only us who needs to pick it up (if that's useful)

- to reduce the time spent by dotcom on managing their infrastructure
- to standardise process: _best practice by default_

## Preparation

### Turn the `dotcom-rendering` repo into a single `yarn` workspace

> the `dotcom-rendering` repo currently managed by two different package managers:

Reviewer: ```suggestion
the `dotcom-rendering` repo is currently managed by two different package managers:
```

1. A `yarn` workspace containing `dotcom-rendering` and `common-rendering`

> 2. An NPM-managed package that contains the `apps-rendering` code.

Reviewer: ```suggestion
2. An `npm`-managed package that contains the `apps-rendering` code.
```

> _n.b. `common-rendering` is only consumed by AR._

Reviewer 2: Not entirely true, the live blogs are shared (`LiveLayout.tsx`, `KeyEventsContainer.tsx`, `LiveBlock.tsx`)

Reviewer 3: I think most of `common-rendering` is actually used by DCR. The components that Max mentioned ^, but also `Accordion`, `Pagination` and the fixtures. There are only couple of things that are exclusively used by AR, like the `FigCaption` component and parts of the `editorialPalette` modules.

Reviewer: that's good to know, thanks very much

> _n.b. `common-rendering` is only consumed by AR._

Reviewer: ```suggestion
```

#### Problem

- `csnx` uses Nx to orchestrate the project
- we want the actual importing of `dotcom-rendering` code to require as few changes as possible

#### Solution

- we will add Nx to the existing the `dotcom-rendering` repo monorepo
- the setup will be as close as possible to `csnx`.

_n.b. `apps-rendering` and `common-rendering` will both be able to use Nx, although setting this up will not be CSTI's initial focus._

### Remove unused config from the `dotcom-rendering` repo

#### Problem

- there is a lot of config files in the `dotcom-rendering` repo
- we don't know how many are used still (e.g. `arkit.json`?)

#### Solution

- identify and remove unused config files

### Align dev dependency versions in the `dotcom-rendering` repo with `csnx`

#### Problem

- `csnx` tries to use a single version of dev dependencies (e.g. Jest, Storybook, ESLint, Prettier etc)
- this makes it easier to manage updates across projects

#### Solution

- align `dotcom-rendering` dev dependencies with `csnx`

### Align Node version in the `dotcom-rendering` repo with `csnx`

#### Problem

- projects in a monorepo cannot use different versions of Node

#### Solution

> - align the `dotcom-rendering` repo `.nvmrc` with `csnx`

Reviewer: i guess either that or changing csnx. but if it means changing this repo then absolutely we'll need a bit of assistance. as with all these things, we'll propose the change and work with teams on it to make sure everyone who needs/wants to be is in alignment

### Align the Storybook setup in the `dotcom-rendering` repo with `csnx`

#### Problem

- `csnx` has a per-project Storybook setup, with a root Storybook that composes them all
- the `dotcom-rendering` repo has a single Storybook that contains stories from all projects

#### Solution

- set up per-project storybooks in the `dotcom-rendering` repo

> - standardise on idiomatic storybook code/structure (if needed)

Reviewer 3: This sounds like a good idea 👍 , would be interested to hear more about how this will look!

## Questions

### What happens to `apps-rendering` and `common-rendering`?

> `apps-rendering` and `common-rendering` will remain in the `dotcom-rendering` repo for now.

Reviewer 3: Given that `common-rendering` is used by DCR, is the plan to a) copy everything that's used into DCR, or b) publish it as a package? If the answer is a) would it make sense to do the same in AR (i.e. copy everything across)? If it's no longer being shared there probably isn't a need for the complexity of managing it as a separate sub-project?

Reviewer: as we mentioned yesterday, this is probably something for @guardian/dotcom-apps-collab to decide, but we can certainly help with whatever the teams think is the best option

### Whats processes/services/workflows do `dotcom-rendering` and `apps-rendering` currently share?

For example:

- github actions
- Riff raff

> - Cypress

Reviewer 3: - I think most of the GH actions are separate, apart from the ones relating to storybook/Chromatic
- Dependabot config is shared
- RiffRaff is separate
- TeamCity is separate, although both TC builds are triggered on PR/merge
- Cypress isn't used by AR

# Pre-relocation

## Rationalising NPM dependencies across the existing DCR / AR monorepo

The `dotcom-rendering` repo currently contains one yarn workspace (containing DCR and common rendering) and a third, unattached package (AR), whose dependencies are handled using NPM. 

AR consumes some undeclared dependencies from the root yarn workspace.

Common-rendering is only consumed by AR.

We will move all projects into one `yarn` workspace, and resolve any issues around shared dependencies _before_ any migration takes place.

## Adding NX to the workspace for use in DCR

We use NX in CSNX to handle our build processes. Pre-relocation we intend to add NX to the `dotcom-rendering` monorepo.

The setup should be as close as possible to CSNX.

## Identify unused config files in AR/DCR

There are a number of potentially unused files in the old monorepo root folders, are we able to remove these? `arkit.json` is a possible example

## Match package versions to infrastructure deps in CSNX

- Jest, Storybook, ESLint, Prettier etc
- Node versions

## Setup Storybook to match CSNX

- per-project storybooks
- standardise on idiomatic storybook code/structure (if needed)

> ## What happens to AR and common rendering?
> 
> Stays where it is?
> How do we not break it?

Reviewer: ```suggestion
## What happens to AR and `common-rendering`?

- we will set up `dotcom-rendering` repo as full monorepo
- AR and CR will both be able to use NX, although this will not be CSTI's initial focus
- AR and CR will remain in the `dotcom-rendering` repo for now
```

# Moving DCR into the CSNX monorepo

## Context

We have recently been in discussion with the Dotcom platform team regarding the relocation of DCR into the CSNX monorepo.

## What do we hope to achieve?

By relocating it in CSNX, we hope to reduce friction between DCR and the Guardian packages it consumes (Source, Libs etc) by rationalising dependencies and standardising package versions.

# Pre-relocation

## Rationalising NPM dependencies across the existing DCR / AR monorepo

> The `dotcom-rendering` repo currently contains a yarn workspace containing DCR and common rendering and AR, whose dependencies are handled using NPM. AR consumes some undeclared dependencies from the root yarn workspace.

Reviewer: ```suggestion
The `dotcom-rendering` repo currently contains one yarn workspace (containing DCR and common rendering) and a third, unattached package (AR), whose dependencies are handled using NPM. 

AR consumes some undeclared dependencies from the root yarn workspace.
```

Common-rendering is only consumed by AR.

We will move all projects into one `yarn` workspace, and resolve any issues around shared dependencies _before_ any migration takes place.

## Adding NX to the workspace for use in DCR

We use NX in CSNX to handle our build processes. Pre-relocation we intend to add NX to the `dotcom-rendering` monorepo.

The setup should be as close as possible to CSNX.

## Identify unused config files in AR/DCR

There are a number of potentially unused files in the old monorepo root folders, are we able to remove these? `arkit.json` is a possible example

## Match package versions to infrastructure deps in CSNX

- Jest, Storybook, ESLint, Prettier etc
- Node versions

## Setup Storybook to match CSNX

- per-project storybooks
- standardise on idiomatic storybook code/structure (if needed)

## What happens to AR and `common-rendering`?

- we will set up `dotcom-rendering` repo as full monorepo
- AR and CR will both be able to use NX, although this will not be CSTI's initial focus
- AR and CR will remain in the `dotcom-rendering` repo for now

## Whats in the root that each project is using?

Eg how do we handle workflows
Riff raff
Cypress? (DCR only)

> 
> ## What is the end goal?
> 
> Run an automated mirror version of DCR in CSNX
> Stagger switch within Frontend (eg 10% of requests go to CSNX instance of DCR)

Reviewer: ```suggestion
```

> ## Whats in the root that each project is using?
> 
> Eg how do we handle workflows
> Riff raff
> Cypress? (DCR only)
> 
> ## What is the end goal?
> 
> Run an automated mirror version of DCR in CSNX
> Stagger switch within Frontend (eg 10% of requests go to CSNX instance of DCR)

Reviewer: ```suggestion
# Questions

## Whats in the current repo root that each project is using?

- how do we handle workflows
- Riff raff
- Cypress? (DCR only)

## How do manage the final switch over?

- run a mirror of DCR in CSNX that automatically keeps in sync?
- stagger the switch over from inside Frontend (eg 10% > 50% > 100% of requests go to CSNX instance of DCR)
```

- reduce the time spent by dotcom on managing their infrastructure
- standardise process: _best practice by default_

# Pre-relocation

## Rationalising NPM dependencies across the existing DCR / AR monorepo

The `dotcom-rendering` repo currently contains one yarn workspace (containing DCR and common rendering) and a third, unattached package (AR), whose dependencies are handled using NPM. 

AR consumes some undeclared dependencies from the root yarn workspace.

Common-rendering is only consumed by AR.

We will move all projects into one `yarn` workspace, and resolve any issues around shared dependencies _before_ any migration takes place.

## Adding NX to the workspace for use in DCR

We use NX in CSNX to handle our build processes. Pre-relocation we intend to add NX to the `dotcom-rendering` monorepo.

The setup should be as close as possible to CSNX.

## Identify unused config files in AR/DCR

There are a number of potentially unused files in the old monorepo root folders, are we able to remove these? `arkit.json` is a possible example

## Match package versions to infrastructure deps in CSNX

- Jest, Storybook, ESLint, Prettier etc
- Node versions

> ## Setup Storybook to match CSNX
> 
> Per project storybooks
> Do we need a composed storybook for AR/DCR/common?
> Storybooks.gutools.co.uk ?
> Add to composed storybook
> Remove idiosyncratic code

Reviewer: ```suggestion
## Setup Storybook to match CSNX

- per-project storybooks
```

# Moving DCR into the CSNX monorepo

## Context

We want to relocate DCR in the CSNX monorepo.

## What do we hope to achieve?

- make it easier for Guardian product teams to ship by reducing the friction of interdependency between DCR and other Guardian projects (Source, Libs, consent, commercial code etc) 
- reduce the time spent by dotcom on managing their infrastructure
- standardise process: _best practice by default_

# Pre-relocation

## Rationalising NPM dependencies across the existing DCR / AR monorepo

The `dotcom-rendering` repo currently contains one yarn workspace (containing DCR and common rendering) and a third, unattached package (AR), whose dependencies are handled using NPM. 

AR consumes some undeclared dependencies from the root yarn workspace.

Common-rendering is only consumed by AR.

> We will move all projects into one `yarn` workspace, and resolve any issues around shared dependencies _before_ any migration takes place.

Reviewer 4: So to confirm, AR will move from NPM into yarn worskpaces as a workspace/package? or they will all be in one shared worskspace/package?

Author: DCR / AR / CR all as packages in one shared Yarn workspace

## Adding NX to the workspace for use in DCR

> We use NX in CSNX to handle our build processes. Pre-relocation we intend to add NX to the `dotcom-rendering` monorepo.

Reviewer 4: Will this be for AR as well as DCR?

Reviewer 4: (seen this is addressed below)

The setup should be as close as possible to CSNX.

## Identify unused config files in AR/DCR

There are a number of potentially unused files in the old monorepo root folders, are we able to remove these? `arkit.json` is a possible example

## Match package versions to infrastructure deps in CSNX

- Jest, Storybook, ESLint, Prettier etc
- Node versions

## Setup Storybook to match CSNX

- per-project storybooks
- standardise on idiomatic storybook code/structure (if needed)

## What happens to AR and `common-rendering`?

- we will set up `dotcom-rendering` repo as full monorepo
- AR and CR will both be able to use NX, although this will not be CSTI's initial focus
- AR and CR will remain in the `dotcom-rendering` repo for now

> # Questions

Reviewer 4: I noticed we haven't mentioned moving the dcr repo to PNPM? Will this be necessary pre-migration?

Author: CSNX is PNPM but adding it to DCR at this stage will add further complexity that we may be able to avoid so it's TBC

> ## Whats processes/services do DCR and AR currently share?

Reviewer 4: Our only shared 'action' is storybook I believe, as well as some packages of course

Reviewer 4: Currently github actions have 'ignore' properties set so DCR actions ignore the AR directory, and vice-versa

Author: Super, thanks!
